src/EditorFeatures/Core/Extensibility/NavigationBar/AbstractEditorNavigationBarItemService.cs

```diff
@@ -23,44 +23,55 @@ protected AbstractEditorNavigationBarItemService(IThreadingContext threadingCont
+            var workspace = document.Project.Solution.Workspace;
 
-            var compilation = await document.Project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
-            var symbolInfo = item.NavigationSymbolId.Resolve(compilation, ignoreAssemblyKey: true, cancellationToken: cancellationToken);
-            var symbol = symbolInfo.GetAnySymbol();
+            var (documentId, position, virtualSpace) = await GetNavigationLocationAsync(document, item, symbolItem, textSnapshot, cancellationToken).ConfigureAwait(false);
+            var navigationService = workspace.Services.GetRequiredService<IDocumentNavigationService>();
 
-            // Do not allow third party navigation to types or constructors
-            if (symbol != null &&
-                symbol is not ITypeSymbol &&
-                !symbol.IsConstructor() &&
-                await symbolNavigationService.TrySymbolNavigationNotifyAsync(symbol, document.Project, cancellationToken).ConfigureAwait(false))
+            if (navigationService.CanNavigateToPosition(workspace, documentId, position, virtualSpace, cancellationToken))
+            {
+                navigationService.TryNavigateToPosition(workspace, documentId, position, virtualSpace, options: null, cancellationToken);
+            }
+            else
             {
-                return;
+                var notificationService = workspace.Services.GetRequiredService<INotificationService>();
+                notificationService.SendNotification(EditorFeaturesResources.The_definition_of_the_object_is_hidden, severity: NotificationSeverity.Error);
             }
+        }
 
-            var navigationPoint = await this.GetSymbolItemNavigationPointAsync(document, item, cancellationToken).ConfigureAwait(false);
-            if (navigationPoint.HasValue)
+        internal virtual Task<(DocumentId documentId, int position, int virtualSpace)> GetNavigationLocationAsync(
+            Document document,
+            NavigationBarItem item,
+            RoslynNavigationBarItem.SymbolItem symbolItem,
+            ITextSnapshot textSnapshot,
+            CancellationToken cancellationToken)
+        {
+            if (item.NavigationTrackingSpan != null)
             {
-                await this.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-                NavigateToVirtualTreePoint(document.Project.Solution, navigationPoint.Value, cancellationToken);
+                return Task.FromResult((document.Id, item.NavigationTrackingSpan.GetSpan(textSnapshot).Start.Position, 0));
+            }
+            else
+            {
+                Contract.ThrowIfNull(symbolItem.Location.OtherDocumentInfo);
+                var otherLocation = symbolItem.Location.OtherDocumentInfo.Value;
+                return Task.FromResult((otherLocation.documentId, otherLocation.navigationSpan.Start, 0));
             }
```

Comment:
by default, all symbol navigation can be driven entirely based on the information stored in the item.